File: leetcode/problems/cpp/binary-tree-level-order-traversal.cpp
// by @codeAbinash
// Time : O(n)
// Space : O(n)

#include "queue"
#include "vector"
using namespace std;

struct TreeNode {
   int val;
   TreeNode* left;
   TreeNode* right;
   TreeNode() : val(0), left(nullptr), right(nullptr) {}
   TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
   TreeNode(int x, TreeNode* left, TreeNode* right) : val(x), left(left), right(right) {}
};

class Solution {
  public:
   vector<vector<int>> levelOrder(TreeNode* root) {
      vector<vector<int>> res;
      if (!root) return res;
      queue<TreeNode*> q;
      q.push(root);
      while (!q.empty()) {
         int size = q.size();
         vector<int> level;
         while (size--) {
            TreeNode* node = q.front();
            q.pop();
            level.push_back(node->val);
            if (node->left) q.push(node->left);
            if (node->right) q.push(node->right);
         }
         res.push_back(level);
      }
      return res;
   }
};

File: leetcode/problems/java/binary-tree-level-order-traversal.java
import java.util.List;
import java.util.ArrayList;

class Solution {
  public List<List<Integer>> levelOrder(TreeNode root) {
    List<List<Integer>> levels = new ArrayList<>();
    if (root == null)
      return levels;
    List<TreeNode> level = new ArrayList<>();
    level.add(root);
    while (!level.isEmpty()) {
      List<TreeNode> nextLevel = new ArrayList<>();
      List<Integer> values = new ArrayList<>();
      for (TreeNode node : level) {
        values.add(node.val);
        if (node.left != null)
          nextLevel.add(node.left);
        if (node.right != null)
          nextLevel.add(node.right);
      }
      levels.add(values);
      level = nextLevel;
    }
    return levels;
  }
}